### PR TITLE
[10843] Correctly protect endpoints collections

### LINKED ISSUE #1819: Segmentation fault in RTPSDomainImpl::find_local_reader [10842]
# Segmentation fault in RTPSDomainImpl::find_local_reader

Callstack: 

```
Stack trace (most recent call last) in thread 29:
#10   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
#9    Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7f4a124874ce, in clone
#8    Object "/lib/x86_64-linux-gnu/libpthread.so.0", at 0x7f4a128bafa2, in 
#7    Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6", at 0x7f4a127a7b2e, in 
#6    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12b64a26, in eprosima::fastrtps::rtps::ResourceEvent::event_service()
#5    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12b6452b, in eprosima::fastrtps::rtps::ResourceEvent::do_timer_actions()
#4    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12b65a3f, in eprosima::fastrtps::rtps::TimedEventImpl::trigger(std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)
#3    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12b7ca64, in 
#2    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12b72bc3, in eprosima::fastrtps::rtps::StatefulWriter::intraprocess_heartbeat(eprosima::fastrtps::rtps::ReaderProxy*, bool)
#1    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12bff2cb, in eprosima::fastrtps::rtps::RTPSDomainImpl::find_local_reader(eprosima::fastrtps::rtps::GUID_t const&)
#0    Object "/usr/local/lib/libfastrtps.so.0", at 0x7f4a12bf4fb7, in eprosima::fastrtps::rtps::RTPSParticipantImpl::find_local_reader(eprosima::fastrtps::rtps::GUID_t const&)
Segmentation fault (Address not mapped to object [0x10])
``` 

RTPSDomainImpl::find_local_reader does not acquire the participant mutex lock.
`m_allReaderList` may be modified by other contexts while processing the for loop. 

```c++
RTPSReader* RTPSParticipantImpl::find_local_reader(
        const GUID_t& reader_guid)
{
    // As this is only called from RTPSDomainImpl::find_local_reader, and it has
    // the domain mutex taken, there is no need to take the participant mutex
    // std::lock_guard<std::recursive_mutex> guard(*mp_mutex);
    for (auto reader : m_allReaderList)
    {
        if (reader->getGuid() == reader_guid)
        {
            return reader;
        }
    }

 

    return nullptr;
}
```

Does the domain mutex protect `m_allReaderList`?

## Expected Behavior

No crash

## Current Behavior

Rarely crash

## Steps to Reproduce

1. Run multiple participants and read/write some data.
2. Heartbeat event and matching event occur at the same time?

Sorry, I'm not sure.

## System information

Fast-RTPS version: 2.2.0  
OS: Debian 10

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
This PR closes #1819. It also replaces #1823.

I have tried the regression test proposed in that PR but it does not seem to provoke the data race condition and the segmentation fault. However, the test is prone to failure due to discovery issues. Thus, I have not introduced it in this PR (if the reviewer considers it is of interest I can update this PR). There is no regression test because forcing the failure is not straight forward nor easy. However, the benefits of correctly protecting the access to the collections are too great compare to a regression test. My opinion is that this fix should be merged.

### CODE REVIEW

**Comment by JLBuenoLopez-eProsima** [comment]
CI failures seem unrelated

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 1]
```diff
@@ -493,7 +493,7 @@ bool StatefulWriter::intraprocess_heartbeat(
     bool returned_value = false;
 
     std::lock_guard<RecursiveTimedMutex> guardW(mp_mutex);
-    RTPSReader* reader = RTPSDomainImpl::find_local_reader(reader_proxy->guid());
+    RTPSReader* reader = reader_proxy->local_reader();
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp:496]: I think this should be left untouched.

There is a corner case where the proxy could still have a pointer to an already deleted reader.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Pleas test aarch64

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test aarch64

**Pull request merged**